
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>xenomapper2 &#8212; xenomapper2 2.0.0a1 documentation</title>
    <link rel="stylesheet" href="_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/custom.css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">xenomapper2 2.0.0a1 documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="xenomapper2">
<h1>xenomapper2<a class="headerlink" href="#xenomapper2" title="Permalink to this headline">¶</a></h1>
<div class="section" id="xenomapper2-xenomapper2-module">
<h2>xenomapper2.xenomapper2 module<a class="headerlink" href="#xenomapper2-xenomapper2-module" title="Permalink to this headline">¶</a></h2>
<span class="target" id="module-xenomapper2.xenomapper2"></span><dl class="class">
<dt id="xenomapper2.xenomapper2.AlignbatchFileReader">
<em class="property">class </em><code class="sig-prename descclassname">xenomapper2.xenomapper2.</code><code class="sig-name descname">AlignbatchFileReader</code><span class="sig-paren">(</span><em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#xenomapper2.xenomapper2.AlignbatchFileReader" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">pylazybam.bam.FileReader</span></code></p>
<p>A BAM file reader that iterates batches of reads with the same name</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>ubam</strong> (<em>BinaryIO</em>) – An binary (bytes) file or stream containing a valid uncompressed
bam file conforming to the specification.</p>
</dd>
<dt class="field-even">Yields</dt>
<dd class="field-even"><p><strong>alignbatch</strong> (<em>List[bytes]</em>) – A byte string of a bam alignment entry in raw binary format</p>
</dd>
</dl>
<dl class="attribute">
<dt id="xenomapper2.xenomapper2.AlignbatchFileReader.raw_header">
<code class="sig-name descname">raw_header</code><a class="headerlink" href="#xenomapper2.xenomapper2.AlignbatchFileReader.raw_header" title="Permalink to this definition">¶</a></dt>
<dd><p>The raw bytestring representing the bam header</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#bytes" title="(in Python v3.8)">bytes</a></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="xenomapper2.xenomapper2.AlignbatchFileReader.raw_refs">
<code class="sig-name descname">raw_refs</code><a class="headerlink" href="#xenomapper2.xenomapper2.AlignbatchFileReader.raw_refs" title="Permalink to this definition">¶</a></dt>
<dd><p>The raw bytestring representing the reference sequences</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#bytes" title="(in Python v3.8)">bytes</a></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="xenomapper2.xenomapper2.AlignbatchFileReader.header">
<code class="sig-name descname">header</code><a class="headerlink" href="#xenomapper2.xenomapper2.AlignbatchFileReader.header" title="Permalink to this definition">¶</a></dt>
<dd><p>The ASCII representation of the header</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="xenomapper2.xenomapper2.AlignbatchFileReader.refs">
<code class="sig-name descname">refs</code><a class="headerlink" href="#xenomapper2.xenomapper2.AlignbatchFileReader.refs" title="Permalink to this definition">¶</a></dt>
<dd><p>A dictionary of reference_name keys with reference_length</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>Dict[str:int]</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="xenomapper2.xenomapper2.AlignbatchFileReader.ref_to_index">
<code class="sig-name descname">ref_to_index</code><a class="headerlink" href="#xenomapper2.xenomapper2.AlignbatchFileReader.ref_to_index" title="Permalink to this definition">¶</a></dt>
<dd><p>A dictionary mapping reference names to the bam numeric identifier</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>Dict[str:int]</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="xenomapper2.xenomapper2.AlignbatchFileReader.index_to_ref">
<code class="sig-name descname">index_to_ref</code><a class="headerlink" href="#xenomapper2.xenomapper2.AlignbatchFileReader.index_to_ref" title="Permalink to this definition">¶</a></dt>
<dd><p>A dictionary mapping bam reference numeric identifiers to names</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>Dict[int:str]</p>
</dd>
</dl>
</dd></dl>

<p class="rubric">Notes</p>
<p>It is advisable not to call the private functions or operate directly
on the underlying file object.</p>
<p>The detailed specification for the BAM format can be found at
<a class="reference external" href="https://samtools.github.io/hts-specs/SAMv1.pdf">https://samtools.github.io/hts-specs/SAMv1.pdf</a></p>
<p class="rubric">Example</p>
<p>This class requires an uncompressed bam file as input.
For this example we will use gzip to decompress the test file included
as a resource in the package</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">gzip</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pkg_resources</span> <span class="kn">import</span> <span class="n">resource_stream</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ubam</span> <span class="o">=</span> <span class="n">gzip</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s1">&#39;/tests/data/paired_end_testdata_human.bam&#39;</span><span class="p">))</span>
</pre></div>
</div>
<p>A bam filereader object can then be created and headers inspected</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">mybam</span> <span class="o">=</span> <span class="n">bam</span><span class="o">.</span><span class="n">FileReader</span><span class="p">(</span><span class="n">ubam</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">mybam</span><span class="o">.</span><span class="n">header</span><span class="p">)</span>
</pre></div>
</div>
<p>The filereader object is an iterator and yields batches of alignments
in raw BAM binary format</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">align_batch</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">mybam</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">align</span> <span class="ow">in</span> <span class="n">align_batch</span><span class="p">:</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="nb">print</span><span class="p">(</span><span class="n">mybam</span><span class="o">.</span><span class="n">index_to_ref</span><span class="p">[</span><span class="n">get_ref_index</span><span class="p">(</span><span class="n">align</span><span class="p">)],</span>
<span class="gp">&gt;&gt;&gt; </span>          <span class="n">get_pos</span><span class="p">(</span><span class="n">align</span><span class="p">),</span>
<span class="gp">&gt;&gt;&gt; </span>          <span class="n">get_AS</span><span class="p">(</span><span class="n">align</span><span class="p">))</span>
</pre></div>
</div>
<dl class="method">
<dt id="xenomapper2.xenomapper2.AlignbatchFileReader.close">
<code class="sig-name descname">close</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#xenomapper2.xenomapper2.AlignbatchFileReader.close" title="Permalink to this definition">¶</a></dt>
<dd><p>Close input file</p>
</dd></dl>

<dl class="method">
<dt id="xenomapper2.xenomapper2.AlignbatchFileReader.get_full_raw_header">
<code class="sig-name descname">get_full_raw_header</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#xenomapper2.xenomapper2.AlignbatchFileReader.get_full_raw_header" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a complete BAM header with all elements suitable for writing to a
bam.FileWriter</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The complete raw BAM header including the BAM magic and reference
block</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#bytes" title="(in Python v3.8)">bytes</a></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="xenomapper2.xenomapper2.AlignbatchFileReader.get_updated_header">
<code class="sig-name descname">get_updated_header</code><span class="sig-paren">(</span><em class="sig-param">id: str</em>, <em class="sig-param">program: str</em>, <em class="sig-param">version: str</em>, <em class="sig-param">command: str = None</em>, <em class="sig-param">description: str = None</em>, <em class="sig-param">raw_header=None</em><span class="sig-paren">)</span><a class="headerlink" href="#xenomapper2.xenomapper2.AlignbatchFileReader.get_updated_header" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a modified version of the header with additional program information
in a &#64;PG BAM header suitable for writing to a new output file.</p>
<p>Does not include the BAM magic or the BAM reference block.</p>
<p>Note that the header should be written after self.magic and before
self.raw_refs</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>id</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – a unique identifier of this action of the program on the BAM</p></li>
<li><p><strong>program</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – the name of the program used to process the BAM</p></li>
<li><p><strong>version</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – version number of the program used to process the BAM</p></li>
<li><p><strong>command</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – the command and arguments to the program used to process the BAM</p></li>
<li><p><strong>description</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – optional description</p></li>
<li><p><strong>raw_header</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#bytes" title="(in Python v3.8)"><em>bytes</em></a>) – a raw BAM formatted header without BAM magic or REF block
used for recursive modification to add multiple</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Parameters will be utf-8 encoded
If there are existing &#64;PG records the last record will be used as PP</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#xenomapper2.xenomapper2.AlignbatchFileReader.update_header" title="xenomapper2.xenomapper2.AlignbatchFileReader.update_header"><code class="xref py py-meth docutils literal notranslate"><span class="pre">update_header()</span></code></a>, <a class="reference internal" href="#xenomapper2.xenomapper2.AlignbatchFileReader.get_full_raw_header" title="xenomapper2.xenomapper2.AlignbatchFileReader.get_full_raw_header"><code class="xref py py-meth docutils literal notranslate"><span class="pre">get_full_raw_header()</span></code></a></p>
</div>
</dd></dl>

<dl class="method">
<dt id="xenomapper2.xenomapper2.AlignbatchFileReader.reset_alignments">
<code class="sig-name descname">reset_alignments</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#xenomapper2.xenomapper2.AlignbatchFileReader.reset_alignments" title="Permalink to this definition">¶</a></dt>
<dd><p>Reset the file pointer to the beginning of the alignment block</p>
</dd></dl>

<dl class="method">
<dt id="xenomapper2.xenomapper2.AlignbatchFileReader.update_header">
<code class="sig-name descname">update_header</code><span class="sig-paren">(</span><em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#xenomapper2.xenomapper2.AlignbatchFileReader.update_header" title="Permalink to this definition">¶</a></dt>
<dd><p>Use get_updated_header to update self.raw_header inplace</p>
<p>See get_updated_header for Parameters and documentation</p>
</dd></dl>

<dl class="method">
<dt id="xenomapper2.xenomapper2.AlignbatchFileReader.update_header_length">
<code class="sig-name descname">update_header_length</code><span class="sig-paren">(</span><em class="sig-param">raw_header=None</em><span class="sig-paren">)</span><a class="headerlink" href="#xenomapper2.xenomapper2.AlignbatchFileReader.update_header_length" title="Permalink to this definition">¶</a></dt>
<dd><p>Update the length of the SAM format text component of the header</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>raw_header</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#bytes" title="(in Python v3.8)"><em>bytes</em></a>) – optional raw SAM text header as bytes to process
Default self.raw_header</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>returns the length corrected raw header if a raw header is provided</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#bytes" title="(in Python v3.8)">bytes</a></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="xenomapper2.xenomapper2.DummyFile">
<em class="property">class </em><code class="sig-prename descclassname">xenomapper2.xenomapper2.</code><code class="sig-name descname">DummyFile</code><span class="sig-paren">(</span><em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#xenomapper2.xenomapper2.DummyFile" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.8)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>A dummy io class - looks like pylazybam.bam.FileWriter but does nothing</p>
<dl class="method">
<dt id="xenomapper2.xenomapper2.DummyFile.close">
<code class="sig-name descname">close</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#xenomapper2.xenomapper2.DummyFile.close" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="xenomapper2.xenomapper2.DummyFile.get_updated_header">
<code class="sig-name descname">get_updated_header</code><span class="sig-paren">(</span><em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#xenomapper2.xenomapper2.DummyFile.get_updated_header" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="xenomapper2.xenomapper2.DummyFile.seekable">
<code class="sig-name descname">seekable</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#xenomapper2.xenomapper2.DummyFile.seekable" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="xenomapper2.xenomapper2.DummyFile.update_header">
<code class="sig-name descname">update_header</code><span class="sig-paren">(</span><em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#xenomapper2.xenomapper2.DummyFile.update_header" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="xenomapper2.xenomapper2.DummyFile.update_header_length">
<code class="sig-name descname">update_header_length</code><span class="sig-paren">(</span><em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#xenomapper2.xenomapper2.DummyFile.update_header_length" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="xenomapper2.xenomapper2.DummyFile.writable">
<code class="sig-name descname">writable</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#xenomapper2.xenomapper2.DummyFile.writable" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="xenomapper2.xenomapper2.DummyFile.write">
<code class="sig-name descname">write</code><span class="sig-paren">(</span><em class="sig-param">data</em><span class="sig-paren">)</span><a class="headerlink" href="#xenomapper2.xenomapper2.DummyFile.write" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="xenomapper2.xenomapper2.DummyFile.write_header">
<code class="sig-name descname">write_header</code><span class="sig-paren">(</span><em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#xenomapper2.xenomapper2.DummyFile.write_header" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="xenomapper2.xenomapper2.XenomapperOutputWriter">
<em class="property">class </em><code class="sig-prename descclassname">xenomapper2.xenomapper2.</code><code class="sig-name descname">XenomapperOutputWriter</code><span class="sig-paren">(</span><em class="sig-param">primary_raw_header: bytes</em>, <em class="sig-param">primary_raw_refs: bytes</em>, <em class="sig-param">secondary_raw_header: bytes</em>, <em class="sig-param">secondary_raw_refs: bytes</em>, <em class="sig-param">primary_specific: Union[str</em>, <em class="sig-param">pathlib.Path</em>, <em class="sig-param">BinaryIO</em>, <em class="sig-param">None] = None</em>, <em class="sig-param">primary_multi: Union[str</em>, <em class="sig-param">pathlib.Path</em>, <em class="sig-param">BinaryIO</em>, <em class="sig-param">None] = None</em>, <em class="sig-param">secondary_specific: Union[str</em>, <em class="sig-param">pathlib.Path</em>, <em class="sig-param">BinaryIO</em>, <em class="sig-param">None] = None</em>, <em class="sig-param">secondary_multi: Union[str</em>, <em class="sig-param">pathlib.Path</em>, <em class="sig-param">BinaryIO</em>, <em class="sig-param">None] = None</em>, <em class="sig-param">unresolved: Union[str</em>, <em class="sig-param">pathlib.Path</em>, <em class="sig-param">BinaryIO</em>, <em class="sig-param">None] = None</em>, <em class="sig-param">unassigned: Union[str</em>, <em class="sig-param">pathlib.Path</em>, <em class="sig-param">BinaryIO</em>, <em class="sig-param">None] = None</em>, <em class="sig-param">basename: str = None</em>, <em class="sig-param">cmdline: str = ''</em>, <em class="sig-param">compresslevel: int = 6</em><span class="sig-paren">)</span><a class="headerlink" href="#xenomapper2.xenomapper2.XenomapperOutputWriter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.8)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Container class grouping xenomapper output files &amp; manipulating headers</p>
<p>Updated headers are written to these files on initialization. The primary
BAM header is used for all files except secondary_specific and
secondary_multi.
A &#64;PO line is added to the file indicating the id and program as xenomapper
with the version, a description field indicating the type of xenomapper
output in the file (eg primary_specific) and the command used to run it.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>primary_raw_header</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#bytes" title="(in Python v3.8)"><em>bytes</em></a>) – the raw header from the primary species BAM file</p></li>
<li><p><strong>primary_raw_refs</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#bytes" title="(in Python v3.8)"><em>bytes</em></a>) – the raw reference sequence info from the primary species BAM file</p></li>
<li><p><strong>secondary_raw_header</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#bytes" title="(in Python v3.8)"><em>bytes</em></a>) – the raw header from the secondary species BAM file</p></li>
<li><p><strong>secondary_raw_refs</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#bytes" title="(in Python v3.8)"><em>bytes</em></a>) – the raw reference sequence info from the secondary species BAM file</p></li>
<li><p><strong>primary_specific</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a><em> or </em><em>Path</em><em> or </em><em>BinaryIO</em><em>, </em><em>optional</em>) – output file for uniquely mapping primary specific alignments
[ Default : None ]</p></li>
<li><p><strong>primary_multi</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a><em> or </em><em>Path</em><em> or </em><em>BinaryIO</em><em>, </em><em>optional</em>) – output file for multimapping primary specific alignments
[ Default : None ]</p></li>
<li><p><strong>secondary_specific</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a><em> or </em><em>Path</em><em> or </em><em>BinaryIO</em><em>, </em><em>optional</em>) – output file for uniquely mapping secondary specific alignments
[ Default : None ]</p></li>
<li><p><strong>secondary_multi</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a><em> or </em><em>Path</em><em> or </em><em>BinaryIO</em><em>, </em><em>optional</em>) – output file for multimapping secondary specific alignments
[ Default : None ]</p></li>
<li><p><strong>unresolved</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a><em> or </em><em>Path</em><em> or </em><em>BinaryIO</em><em>, </em><em>optional</em>) – output file for alignments where primary or secondary is ambiguous
[ Default : None ]</p></li>
<li><p><strong>unassigned</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a><em> or </em><em>Path</em><em> or </em><em>BinaryIO</em><em>, </em><em>optional</em>) – output file for reads that do not map sufficiently well in either BAM
[ Default : None ]</p></li>
<li><p><strong>basename</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a><em>, </em><em>optional</em>) – filename stem to derive all output filenames &lt;basename&gt;_&lt;outfilename&gt;
(eg basename=’Foo’ -&gt; Foo_primary_specific, … Foo_unassigned)
[ Default : None ]</p></li>
<li><p><strong>cmdline</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a><em>, </em><em>optional</em>) – The commandline to include in the output BAM header</p></li>
<li><p><strong>compresslevel</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>, </em><em>optional</em>) – gzip compression level for output file
[ Default : 6 ]</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>A container class with __get_item__ , keys, close</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">xow</span> <span class="o">=</span> <span class="n">XenomapperOutputWriter</span><span class="p">(</span><span class="n">phead</span><span class="p">,</span><span class="n">pref</span><span class="p">,</span><span class="n">shead</span><span class="p">,</span><span class="n">sref</span><span class="p">,</span><span class="n">basename</span><span class="o">=</span><span class="s1">&#39;/foo&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">xow</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">xow</span><span class="p">[</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">xow</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<dl class="method">
<dt id="xenomapper2.xenomapper2.XenomapperOutputWriter.close">
<code class="sig-name descname">close</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#xenomapper2.xenomapper2.XenomapperOutputWriter.close" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="xenomapper2.xenomapper2.XenomapperOutputWriter.keys">
<code class="sig-name descname">keys</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#xenomapper2.xenomapper2.XenomapperOutputWriter.keys" title="Permalink to this definition">¶</a></dt>
<dd><p>return the keys for the file output objects
:returns:
:rtype: Iterable[str]</p>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="xenomapper2.xenomapper2.always_very_negative">
<code class="sig-prename descclassname">xenomapper2.xenomapper2.</code><code class="sig-name descname">always_very_negative</code><span class="sig-paren">(</span><em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#xenomapper2.xenomapper2.always_very_negative" title="Permalink to this definition">¶</a></dt>
<dd><p>A function that returns -2**31 for any combination of parameters
:returns: always -2**31 == MIN32INT == -2147483648
:rtype: int</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#xenomapper2.xenomapper2.always_zero" title="xenomapper2.xenomapper2.always_zero"><code class="xref py py-func docutils literal notranslate"><span class="pre">always_zero()</span></code></a></p>
</div>
</dd></dl>

<dl class="function">
<dt id="xenomapper2.xenomapper2.always_zero">
<code class="sig-prename descclassname">xenomapper2.xenomapper2.</code><code class="sig-name descname">always_zero</code><span class="sig-paren">(</span><em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#xenomapper2.xenomapper2.always_zero" title="Permalink to this definition">¶</a></dt>
<dd><p>A function that returns zero for any combination of parameters
:returns: always 0
:rtype: int</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#xenomapper2.xenomapper2.always_very_negative" title="xenomapper2.xenomapper2.always_very_negative"><code class="xref py py-func docutils literal notranslate"><span class="pre">always_very_negative()</span></code></a></p>
</div>
</dd></dl>

<dl class="function">
<dt id="xenomapper2.xenomapper2.calc_cigar_based_score">
<code class="sig-prename descclassname">xenomapper2.xenomapper2.</code><code class="sig-name descname">calc_cigar_based_score</code><span class="sig-paren">(</span><em class="sig-param">cigar_string: bytes</em>, <em class="sig-param">NM: int = None</em>, <em class="sig-param">mismatch: int = -6</em>, <em class="sig-param">gap_open: int = -5</em>, <em class="sig-param">gap_extend: int = -3</em>, <em class="sig-param">softclip: int = -2</em><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#xenomapper2.xenomapper2.calc_cigar_based_score" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate values for a score equivalent to a
rescaled AS tag using the SAM formatted cigar line and NM tags from an
alignment.
Score is scaled such that a perfect match for the full
length of the read achieves a score of 0.
Standard penalties are: mismatch -6, gap open -5, gap extend -3,
softclipped -2 (equiv to +2 match)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>align</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#bytes" title="(in Python v3.8)"><em>bytes</em></a>) – the raw alignment entry from a BAM file</p></li>
<li><p><strong>mismatch</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – score penalty for a mismatch [ Default : -6 ]</p></li>
<li><p><strong>gap_open</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – score penalty for opening/initiating a gap [ Default : -5 ]</p></li>
<li><p><strong>gap_extend</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – score penalty for extending an existing gap [ Default : -3 ]</p></li>
<li><p><strong>softclip</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – score penalty for removing bases at the start or end of a read
[ Default : -2 ]</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The AS like score calculated from the cigar string
0 maximum and mismatch*seq_length theoretical minimum</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)">int</a></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#xenomapper2.xenomapper2.get_cigar_based_score" title="xenomapper2.xenomapper2.get_cigar_based_score"><code class="xref py py-func docutils literal notranslate"><span class="pre">get_cigar_based_score()</span></code></a></p>
</div>
</dd></dl>

<dl class="function">
<dt id="xenomapper2.xenomapper2.conservative_state_map">
<code class="sig-prename descclassname">xenomapper2.xenomapper2.</code><code class="sig-name descname">conservative_state_map</code><span class="sig-paren">(</span><em class="sig-param">forward_state</em>, <em class="sig-param">reverse_state</em><span class="sig-paren">)</span><a class="headerlink" href="#xenomapper2.xenomapper2.conservative_state_map" title="Permalink to this definition">¶</a></dt>
<dd><p>Conservative tate map for inferring fragment state from forward &amp; reverse
Reads where there is ambiguous evidence will be assigned the least specific
state.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>forward_state</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – a xenomapper state for the forward read
(primary_specific,primary_multi,secondary_specific,secondary_multi,
unassigned, unresolved)</p></li>
<li><p><strong>reverse_state</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a><em> or </em><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – a xenomapper state for the reverse read or None
(primary_specific,primary_multi,secondary_specific,secondary_multi,
unassigned, unresolved) or None</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a xenomapper state for the fragment
(primary_specific,primary_multi,secondary_specific,secondary_multi,
unassigned, unresolved)</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)">str</a></p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>In future these will be of type PEP586 Literal</p>
</dd></dl>

<dl class="function">
<dt id="xenomapper2.xenomapper2.get_bamprimary_AS_XS">
<code class="sig-prename descclassname">xenomapper2.xenomapper2.</code><code class="sig-name descname">get_bamprimary_AS_XS</code><span class="sig-paren">(</span><em class="sig-param">alignments: List[bytes], AS_function=&lt;function get_AS&gt;, XS_function=&lt;function get_XS&gt;</em><span class="sig-paren">)</span> &#x2192; Tuple[int, int]<a class="headerlink" href="#xenomapper2.xenomapper2.get_bamprimary_AS_XS" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the AS and XS of the primary alignment</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>alignments</strong> (<em>List</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#bytes" title="(in Python v3.8)"><em>bytes</em></a><em>]</em>) – a list of binary format BAM alignments from the same read.
Should contain only forward or reverse reads and only one primary flag</p></li>
<li><p><strong>AS_function</strong> – a function that accepts a BAM alignment bytestring and returns the AS
score (alignment score) as an integer
eg pylazybam.bam.get_AS or</p></li>
<li><p><strong>XS_function</strong> – a function that accepts a BAM alignment bytestring and returns the XS
score as an integer</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a tuple of the alignment score AS and suboptimal alignment score XS</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Tuple[<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)">int</a>,<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)">int</a>]</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.8)"><strong>ValueError</strong></a> – raises value error if there is no primary or more than one primary flag</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="xenomapper2.xenomapper2.get_cigar_based_score">
<code class="sig-prename descclassname">xenomapper2.xenomapper2.</code><code class="sig-name descname">get_cigar_based_score</code><span class="sig-paren">(</span><em class="sig-param">align: bytes</em>, <em class="sig-param">no_tag: int = -2147483648</em><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#xenomapper2.xenomapper2.get_cigar_based_score" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a score equivalent to a rescaled AS tag from a raw BAM alignment
Extracts and decodes the cigar string and NM tags and calls
calc_cigar_based_score to calculate</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>align</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#bytes" title="(in Python v3.8)"><em>bytes</em></a>) – A raw alignment from a BAM file</p></li>
<li><p><strong>no_tag</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – The value to return if no NM tag or cigar string is present</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the AS like score with 0 maximum and -6*seq_length theoretical minimum</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)">int</a></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#xenomapper2.xenomapper2.calc_cigar_based_score" title="xenomapper2.xenomapper2.calc_cigar_based_score"><code class="xref py py-func docutils literal notranslate"><span class="pre">calc_cigar_based_score()</span></code></a></p>
</div>
</dd></dl>

<dl class="function">
<dt id="xenomapper2.xenomapper2.get_mapping_state">
<code class="sig-prename descclassname">xenomapper2.xenomapper2.</code><code class="sig-name descname">get_mapping_state</code><span class="sig-paren">(</span><em class="sig-param">AS1: int</em>, <em class="sig-param">XS1: int</em>, <em class="sig-param">AS2: int</em>, <em class="sig-param">XS2: int</em>, <em class="sig-param">min_score: int = -2147483648</em><span class="sig-paren">)</span><a class="headerlink" href="#xenomapper2.xenomapper2.get_mapping_state" title="Permalink to this definition">¶</a></dt>
<dd><p>Determine the mapping state based on scores in each species.
Scores can be negative but better matches must have higher scores</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>AS1</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – score of best match in primary species</p></li>
<li><p><strong>XS1</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – score of suboptimal match in primary species</p></li>
<li><p><strong>AS2</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – score of best match in secondary species</p></li>
<li><p><strong>XS2</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – score of suboptimal match in secondary species</p></li>
<li><p><strong>min_score</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em> [ </em><em>Default : -2**31</em><em> ]</em>) – the score that matches must exceed in order to be
considered valid matches. Note scores equalling this
value will also be considered not to match.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>‘primary_multi’, ‘secondary_multi’,
‘unresolved’, or ‘unassigned’ indicating match state.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>state - a string of ‘primary_specific’, ‘secondary_specific’,</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="xenomapper2.xenomapper2.get_max_AS_XS">
<code class="sig-prename descclassname">xenomapper2.xenomapper2.</code><code class="sig-name descname">get_max_AS_XS</code><span class="sig-paren">(</span><em class="sig-param">alignments: List[bytes], AS_function=&lt;function get_AS&gt;, XS_function=&lt;function get_XS&gt;, sort_key=None, default=-2147483648</em><span class="sig-paren">)</span> &#x2192; Tuple[int, int]<a class="headerlink" href="#xenomapper2.xenomapper2.get_max_AS_XS" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the maximum AS and XS from a group of alignments</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>alignments</strong> (<em>List</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#bytes" title="(in Python v3.8)"><em>bytes</em></a><em>]</em>) – a list of binary format BAM alignments.</p></li>
<li><p><strong>AS_function</strong> (<em>Callable</em><em>[</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#bytes" title="(in Python v3.8)"><em>bytes</em></a><em>]</em><em>, </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>]</em>) – a function that accepts a BAM alignment bytestring and returns the AS
score (alignment score) as an integer
eg pylazybam.bam.get_AS or</p></li>
<li><p><strong>XS_function</strong> (<em>Callable</em><em>[</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#bytes" title="(in Python v3.8)"><em>bytes</em></a><em>]</em><em>, </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>]</em>) – a function that accepts a BAM alignment bytestring and returns the XS
score as an integer</p></li>
<li><p><strong>sort_key</strong> (<em>Callable</em><em>[</em><em>[</em><em>Tuple</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>,</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>]</em><em>]</em><em>, </em><em>Any</em><em>]</em>) – a key function that converts tuples to items to be used by sorted()</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a tuple of the alignment score AS and suboptimal alignment score XS</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Tuple[<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)">int</a>,<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)">int</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="xenomapper2.xenomapper2.output_summary">
<code class="sig-prename descclassname">xenomapper2.xenomapper2.</code><code class="sig-name descname">output_summary</code><span class="sig-paren">(</span><em class="sig-param">category_counts: collections.Counter</em>, <em class="sig-param">title='Read Count Category Summary\n'</em>, <em class="sig-param">outfile=&lt;_io.TextIOWrapper name='&lt;stderr&gt;' mode='w' encoding='UTF-8'&gt;</em><span class="sig-paren">)</span><a class="headerlink" href="#xenomapper2.xenomapper2.output_summary" title="Permalink to this definition">¶</a></dt>
<dd><p>Print a summary table based on a Counter</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>category_counts</strong> (<em>Counter</em>) – </p></li>
<li><p><strong>title</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – </p></li>
<li><p><strong>outfile</strong> (<em>TextIO</em>) – </p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="xenomapper2.xenomapper2.split_forward_reverse">
<code class="sig-prename descclassname">xenomapper2.xenomapper2.</code><code class="sig-name descname">split_forward_reverse</code><span class="sig-paren">(</span><em class="sig-param">alignments: List[bytes]</em><span class="sig-paren">)</span> &#x2192; Tuple[List[bytes], List[bytes]]<a class="headerlink" href="#xenomapper2.xenomapper2.split_forward_reverse" title="Permalink to this definition">¶</a></dt>
<dd><p>Split a list of BAM alignments into forward and reverse</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>alignments</strong> (<em>List</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#bytes" title="(in Python v3.8)"><em>bytes</em></a><em>]</em>) – a list of BAM alignments in raw binary format containing forward
and optionally reverse alignments</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a tuple containing a list of forward and reverse BAM reads</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>(List[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#bytes" title="(in Python v3.8)">bytes</a>], List[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#bytes" title="(in Python v3.8)">bytes</a>])</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.8)"><strong>ValueError</strong></a> – if there are any reads that do not have the forward or reverse flag</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="xenomapper2.xenomapper2.state_map">
<code class="sig-prename descclassname">xenomapper2.xenomapper2.</code><code class="sig-name descname">state_map</code><span class="sig-paren">(</span><em class="sig-param">forward_state: str</em>, <em class="sig-param">reverse_state: str</em><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#xenomapper2.xenomapper2.state_map" title="Permalink to this definition">¶</a></dt>
<dd><p>State map for inferring the fragment state from forward &amp; reverse
Reads where there is stronger but not contradictory evidence will be
assigned the more specific state.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>forward_state</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – a xenomapper state for the forward read
(primary_specific,primary_multi,secondary_specific,secondary_multi,
unassigned, unresolved)</p></li>
<li><p><strong>reverse_state</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a><em> or </em><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.8)"><em>None</em></a>) – a xenomapper state for the reverse read or None
(primary_specific,primary_multi,secondary_specific,secondary_multi,
unassigned, unresolved) or None</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a xenomapper state for the fragment
(primary_specific,primary_multi,secondary_specific,secondary_multi,
unassigned, unresolved)</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)">str</a></p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>In future these will be of type PEP586 Literal</p>
</dd></dl>

<dl class="function">
<dt id="xenomapper2.xenomapper2.xenomap">
<code class="sig-prename descclassname">xenomapper2.xenomapper2.</code><code class="sig-name descname">xenomap</code><span class="sig-paren">(</span><em class="sig-param">primary_bam: xenomapper2.xenomapper2.AlignbatchFileReader</em>, <em class="sig-param">secondary_bam: xenomapper2.xenomapper2.AlignbatchFileReader</em>, <em class="sig-param">output_writer: xenomapper2.xenomapper2.XenomapperOutputWriter</em>, <em class="sig-param">score_function: Callable = &lt;function get_bamprimary_AS_XS&gt;</em>, <em class="sig-param">AS_function: Callable = &lt;function get_AS&gt;</em>, <em class="sig-param">XS_function: Callable = &lt;function get_XS&gt;</em>, <em class="sig-param">min_score: int = -2147483648</em>, <em class="sig-param">conservative: bool = False</em><span class="sig-paren">)</span><a class="headerlink" href="#xenomapper2.xenomapper2.xenomap" title="Permalink to this definition">¶</a></dt>
<dd><p>core method to coordinate the xenomapping of BAMS</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>primary_bam</strong> (<a class="reference internal" href="#xenomapper2.xenomapper2.AlignbatchFileReader" title="xenomapper2.xenomapper2.AlignbatchFileReader"><em>AlignbatchFileReader</em></a>) – An interable that yields iterables of primary BAM alignments</p></li>
<li><p><strong>secondary_bam</strong> (<a class="reference internal" href="#xenomapper2.xenomapper2.AlignbatchFileReader" title="xenomapper2.xenomapper2.AlignbatchFileReader"><em>AlignbatchFileReader</em></a>) – An interable that yields iterables of secondary BAM alignments</p></li>
<li><p><strong>output_writer</strong> (<a class="reference internal" href="#xenomapper2.xenomapper2.XenomapperOutputWriter" title="xenomapper2.xenomapper2.XenomapperOutputWriter"><em>XenomapperOutputWriter</em></a>) – output writer that holds output files of type bam.FileWriter</p></li>
<li><p><strong>score_function</strong> (<em>Callable</em>) – a xenomapper alignment batch calling function
get_bamprimary_AS_XS or get_max_AS_XS</p></li>
<li><p><strong>AS_function</strong> (<em>Callable</em><em>[</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#bytes" title="(in Python v3.8)"><em>bytes</em></a><em>]</em><em>, </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>]</em>) – a function that accepts a BAM alignment bytestring and returns the AS
score (alignment score) as an integer
get_AS or get_cigar_based_score</p></li>
<li><p><strong>XS_function</strong> (<em>Callable</em><em>[</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#bytes" title="(in Python v3.8)"><em>bytes</em></a><em>]</em><em>, </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a><em>]</em>) – a function that accepts a BAM alignment bytestring and returns the XS
score as an integer
get_XS, get_ZS, always_zero, or always_very_negative</p></li>
<li><p><strong>min_score</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – the score that matches must exceed in order to be
considered valid matches. Note scores equalling this
value will also be considered not valid matches.
[ Default : -2**31 ]</p></li>
<li><p><strong>conservative</strong> – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Tuple[Counter, Counter, <a class="reference internal" href="#xenomapper2.xenomapper2.XenomapperOutputWriter" title="xenomapper2.xenomapper2.XenomapperOutputWriter">XenomapperOutputWriter</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="xenomapper2.xenomapper2.xenomap_states">
<code class="sig-prename descclassname">xenomapper2.xenomapper2.</code><code class="sig-name descname">xenomap_states</code><span class="sig-paren">(</span><em class="sig-param">primary_aligns: Iterable[bytes], secondary_aligns: Iterable[bytes], score_function: Callable = &lt;function get_bamprimary_AS_XS&gt;, AS_function: Callable = &lt;function get_AS&gt;, XS_function: Callable = &lt;function get_XS&gt;, min_score: int = -2147483648</em><span class="sig-paren">)</span> &#x2192; Tuple[int, int]<a class="headerlink" href="#xenomapper2.xenomapper2.xenomap_states" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the xenomapping state for the forward and reverse reads</p>
<dl class="simple">
<dt>primary_aligns<span class="classifier">List[bytes]</span></dt><dd><p>a list of binary format BAM alignments from the primary BAM.</p>
</dd>
<dt>secondary_aligns<span class="classifier">List[bytes]</span></dt><dd><p>a list of binary format BAM alignments from the primary BAM.</p>
</dd>
<dt>score_function<span class="classifier">Callable</span></dt><dd><p>a xenomapper alignment batch calling function
get_bamprimary_AS_XS or get_max_AS_XS</p>
</dd>
<dt>AS_function<span class="classifier">Callable[[bytes], int]</span></dt><dd><p>a function that accepts a BAM alignment bytestring and returns the AS
score (alignment score) as an integer
get_AS or get_cigar_based_score</p>
</dd>
<dt>XS_function<span class="classifier">Callable[[bytes], int]</span></dt><dd><p>a function that accepts a BAM alignment bytestring and returns the XS
score as an integer
get_XS, get_ZS, always_zero, or always_very_negative</p>
</dd>
<dt>min_score<span class="classifier">int [ Default</span><span class="classifier">-2**31 ]</span></dt><dd><p>the score that matches must exceed in order to be
considered valid matches. Note scores equalling this
value will also be considered not valid matches.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the xenomapper state of the forward and reverse read</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>Tuple[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)">str</a>,<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)">str</a>]</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="xenomapper2-cli-module">
<h2>xenomapper2.cli module<a class="headerlink" href="#xenomapper2-cli-module" title="Permalink to this headline">¶</a></h2>
<p>Format decoders for BAM alignment data types</p>
<span class="target" id="module-xenomapper2.cli"></span><dl class="function">
<dt id="xenomapper2.cli.main">
<code class="sig-prename descclassname">xenomapper2.cli.</code><code class="sig-name descname">main</code><span class="sig-paren">(</span><em class="sig-param">arguments: str = None</em>, <em class="sig-param">output=&lt;_io.TextIOWrapper name='&lt;stderr&gt;' mode='w' encoding='UTF-8'&gt;</em><span class="sig-paren">)</span> &#x2192; Tuple[collections.Counter, collections.Counter]<a class="headerlink" href="#xenomapper2.cli.main" title="Permalink to this definition">¶</a></dt>
<dd><p>main function for orchestrating a xenomapper2 run</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>arguments</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – a string of unix command line arguments - used for testing</p></li>
<li><p><strong>output</strong> (<em>TextIO</em>) – destination to print progress and results to - used for testing</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>Counter</em> – a counter of pair states</p></li>
<li><p><em>Counter</em> – a counter of read states</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="pylazybam-tags-module">
<h2>pylazybam.tags module<a class="headerlink" href="#pylazybam-tags-module" title="Permalink to this headline">¶</a></h2>
<p>Functions for extracting and decoding SAM tag data from BAM alignments</p>
<span class="target" id="module-pylazybam.tags"></span><dl class="function">
<dt id="pylazybam.tags.get_AS">
<code class="sig-prename descclassname">pylazybam.tags.</code><code class="sig-name descname">get_AS</code><span class="sig-paren">(</span><em class="sig-param">tag_bytes: bytes</em>, <em class="sig-param">no_tag: Any = -2147483648</em><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#pylazybam.tags.get_AS" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract the high scoring alignment score from an AS tag in a raw BAM
alignment bytestring</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tag_bytes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#bytes" title="(in Python v3.8)"><em>bytes</em></a>) – a bytestring containing bam formatted tag elements</p></li>
<li><p><strong>no_tag</strong> (<em>Any</em>) – return value for when tag not found (default: MIN32INT)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>AS Tag Value</strong> – the integer value of the AS tag
returns the value of no_tag if tag absent (default:MIN32INT)</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)">int</a></p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.8)"><strong>ValueError</strong></a> – raises a ValueError if more than one tag match</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Recommended try accept for use on raw alignment with fall back
to calling on only the tag byte string.</p>
<p>Please test carefully on your BAM output as in complicated output the
regular expression based extraction of the tag can be error prone</p>
</dd></dl>

<dl class="function">
<dt id="pylazybam.tags.get_MD">
<code class="sig-prename descclassname">pylazybam.tags.</code><code class="sig-name descname">get_MD</code><span class="sig-paren">(</span><em class="sig-param">tag_bytes: bytes</em>, <em class="sig-param">no_tag: Any = None</em><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#pylazybam.tags.get_MD" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract the MD tag from a raw BAM alignment bytestring</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tag_bytes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#bytes" title="(in Python v3.8)"><em>bytes</em></a>) – a bytestring containing bam formatted tag elements</p></li>
<li><p><strong>no_tag</strong> (<em>Any</em>) – return value for when tag not found (default: None)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>MD Tag Value</strong> – an ASCII string representing the SAM format value of the MD tag
returns the value of no_tag if tag absent (default: None)</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)">str</a></p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.8)"><strong>ValueError</strong></a> – raises a ValueError if more than one tag match</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The MD field aims to achieve SNP/indel calling independent of the reference.
Numbers represent matches, letters bases that differ from the reference and
bases preceeded by ^ are deletions. A ^T0A indicates a base change to an A
immediately following a deleted T.</p>
<p>Recommended try accept for use on raw alignment with fall back
to calling on only the tag byte string.</p>
<p>Please test carefully on your BAM output as in complicated output the
regular expression based extraction of the tag can be error prone</p>
</dd></dl>

<dl class="function">
<dt id="pylazybam.tags.get_XS">
<code class="sig-prename descclassname">pylazybam.tags.</code><code class="sig-name descname">get_XS</code><span class="sig-paren">(</span><em class="sig-param">tag_bytes: bytes</em>, <em class="sig-param">no_tag: Any = -2147483648</em><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#pylazybam.tags.get_XS" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract the suboptimal alignment score from an XS tag in a raw BAM
alignment bytestring</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tag_bytes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#bytes" title="(in Python v3.8)"><em>bytes</em></a>) – a bytestring containing bam formatted tag elements</p></li>
<li><p><strong>no_tag</strong> (<em>Any</em>) – return value for when tag not found (default: MIN32INT)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>XS Tag Value</strong> – the integer value of the XS tag
returns the value of no_tag if tag absent (default:MIN32INT)</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)">int</a></p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.8)"><strong>ValueError</strong></a> – raises a ValueError if more than one tag match</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>This function is for the genome aligner definition of XS where XS:i:&lt;int&gt;
is the alignment score of the suboptimal alignment.
This is not the same as the spliced aligner XS tag XS:C:&lt;str&gt; that
represents the strand on which the intron occurs (equiv to TS:C:&lt;str&gt;)</p>
<p>Recommended try accept for use on raw alignment with fall back
to calling on only the tag byte string.</p>
<p>Please test carefully on your BAM output as in complicated output the
regular expression based extraction of the tag can be error prone</p>
</dd></dl>

<dl class="function">
<dt id="pylazybam.tags.get_ZS">
<code class="sig-prename descclassname">pylazybam.tags.</code><code class="sig-name descname">get_ZS</code><span class="sig-paren">(</span><em class="sig-param">tag_bytes: bytes</em>, <em class="sig-param">no_tag: Any = -2147483648</em><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#pylazybam.tags.get_ZS" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract the suboptimal alignment score from the ZS tag in a raw BAM
alignment bytestring</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tag_bytes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#bytes" title="(in Python v3.8)"><em>bytes</em></a>) – a bytestring containing bam formatted tag elements</p></li>
<li><p><strong>no_tag</strong> (<em>Any</em>) – return value for when tag not found (default: MIN32INT)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>ZS Tag Value</strong> – the integer value of the ZS tag
returns the value of no_tag if tag absent (default:MIN32INT)</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)">int</a></p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.8)"><strong>ValueError</strong></a> – raises a ValueError if more than one tag match</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>ZS is the equivalent to XS:i:&lt;int&gt; tag in some spliced aligners
including HISAT2.</p>
<p>Recommended try accept for use on raw alignment with fall back
to calling on only the tag byte string.</p>
<p>Please test carefully on your BAM output as in complicated output the
regular expression based extraction of the tag can be error prone</p>
</dd></dl>

<dl class="function">
<dt id="pylazybam.tags.get_int_tag">
<code class="sig-prename descclassname">pylazybam.tags.</code><code class="sig-name descname">get_int_tag</code><span class="sig-paren">(</span><em class="sig-param">tag_bytes: bytes</em>, <em class="sig-param">tag: bytes</em>, <em class="sig-param">no_tag: Any = -2147483648</em><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#pylazybam.tags.get_int_tag" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract an integer format tag from a raw BAM alignment bytestring</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tag_bytes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#bytes" title="(in Python v3.8)"><em>bytes</em></a>) – a bytestring containing bam formatted tag elements</p></li>
<li><p><strong>tag</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#bytes" title="(in Python v3.8)"><em>bytes</em></a>) – the two byte tag to be returned eg b’AS’
The tag must represent an integer format tag</p></li>
<li><p><strong>no_tag</strong> (<em>Any</em>) – return value for when tag not found (default: MIN32INT)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the integer value of the tag
returns the value of no_tag if tag absent (default:MIN32INT)</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)">int</a></p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.8)"><strong>ValueError</strong></a> – raises a ValueError if more than one tag match</p>
</dd>
</dl>
<p>Potential values for the tag parameter include:</p>
<blockquote>
<div><dl class="simple">
<dt>AM:i:score  The smallest template-independent mapping quality of any</dt><dd><p>segment in the same template as this read. (See also SM.)</p>
</dd>
</dl>
<p>AS:i:score  Alignment score generated by aligner.</p>
<p>CP:i:pos    Leftmost coordinate of the next hit.</p>
<p>FI:i:int    The index of segment in the template.</p>
<p>H0:i:count  Number of perfect hits.</p>
<p>H1:i:count  Number of 1-difference hits (see also NM).</p>
<p>H2:i:count  Number of 2-difference hits.</p>
<dl class="simple">
<dt>HI:i:i      Query hit index, indicating the alignment record is the</dt><dd><p>i-th one stored in SAM.</p>
</dd>
<dt>IH:i:count  Number of alignments stored in the file that contain the</dt><dd><p>query in the current record.</p>
</dd>
</dl>
<p>MQ:i:score  Mapping quality of the mate/next segment.</p>
<dl class="simple">
<dt>NH:i:count  Number of reported alignments that contain the query in</dt><dd><p>the current record.</p>
</dd>
<dt>NM:i:count  Number of differences (mismatches plus inserted and deleted</dt><dd><p>bases) between the sequence and reference, counting only
(case-insensitive) A, C, G and T bases in sequence and
reference as potential matches, with everything else being
a mismatch.</p>
</dd>
<dt>PQ:i:score  Phred likelihood of the template, conditional on the mapping</dt><dd><p>locations of both/all segments being correct.</p>
</dd>
<dt>SM:i:score  Template-independent mapping quality, i.e., the mapping</dt><dd><p>quality if the read were mapped as a single read rather
than as part of a read pair or template.</p>
</dd>
</dl>
<p>TC:i:       The number of segments in the template.</p>
<dl class="simple">
<dt>UQ:i:       Phred likelihood of the segment,</dt><dd><p>conditional on the mapping being correct.</p>
</dd>
</dl>
<p>see <a class="reference external" href="https://samtools.github.io/hts-specs/SAMtags.pdf">https://samtools.github.io/hts-specs/SAMtags.pdf</a></p>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="pylazybam.tags.get_str_tag">
<code class="sig-prename descclassname">pylazybam.tags.</code><code class="sig-name descname">get_str_tag</code><span class="sig-paren">(</span><em class="sig-param">tag_bytes: bytes</em>, <em class="sig-param">tag: bytes</em>, <em class="sig-param">no_tag: Any = None</em><span class="sig-paren">)</span> &#x2192; str<a class="headerlink" href="#pylazybam.tags.get_str_tag" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract an integer format tag from a raw BAM alignment bytestring</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tag_bytes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#bytes" title="(in Python v3.8)"><em>bytes</em></a>) – a bytestring containing bam formatted tag elements</p></li>
<li><p><strong>tag</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#bytes" title="(in Python v3.8)"><em>bytes</em></a>) – the two byte tag to be returned eg b’AS’
The tag must represent an integer format tag</p></li>
<li><p><strong>no_tag</strong> (<em>Any</em>) – return value for when tag not found (default: None)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>MD Tag Value</strong> – an ASCII string representing the SAM format value of the MD tag
returns the value of no_tag if tag absent (default: None)</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)">str</a></p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.8)"><strong>ValueError</strong></a> – raises a ValueError if more than one tag match</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Potential values for the tag parameter include:</p>
<blockquote>
<div><dl class="simple">
<dt>BQ:Z:qualities  Offset to base alignment quality (BAQ), of the same</dt><dd><p>length as the read sequence. At the i-th read base,
BAQi = Qi − (BQi − 64) where Qi is the i-th base quality</p>
</dd>
</dl>
<p>CC:Z:rname      Reference name of the next hit; ‘=’ for same chromosome.</p>
<dl class="simple">
<dt>E2:Z:bases      The 2nd most likely base calls.</dt><dd><p>Same encoding and same length as SEQ.</p>
</dd>
</dl>
<p>FS:Z:str        Segment suffix.</p>
<p>MC:Z:cigar      CIGAR string for mate/next segment.</p>
<p>MD:Z:           String for mismatching positions.</p>
<dl class="simple">
<dt>Q2:Z:qualities  Phred quality of the mate/next segment sequence in the</dt><dd><p>R2 tag. Same encoding as QUAL.</p>
</dd>
</dl>
<p>R2:Z:bases      Sequence of the mate/next segment in the template.</p>
<dl class="simple">
<dt>SA:Z:           (rname ,pos ,strand ,CIGAR ,mapQ ,NM ;)+</dt><dd><p>Other canonical alignments in a chimeric alignment,
formatted as a semicolon-delimited list.</p>
</dd>
<dt>U2:Z:           Phred probability of the 2nd call being wrong</dt><dd><p>conditional on the best being wrong.</p>
</dd>
</dl>
<p>RG:Z:readgroup  The read group to which the read belongs.</p>
<p>LB:Z:library    The library from which the read has been sequenced.</p>
<p>PG:Z:program id Program. Value matches the header PG-ID tag</p>
<p>PU:Z:platformunit The platform unit in which the read was sequenced.</p>
<p>CO:Z:text       Free-text comments.</p>
<dl class="simple">
<dt>BC:Z:sequence   Barcode sequence (Identifying the sample/library),</dt><dd><p>with any quality scores (optionally) stored in  QT tag.
The BC tag should match the QT tag in length.</p>
</dd>
<dt>QT:Z:qualities  Phred quality of the sample barcode sequence in BC tag.</dt><dd><p>Same encoding as QUAL, i.e., Phred score + 33.</p>
</dd>
<dt>CB:Z:str        Cell identifier, consisting of the optionally-corrected</dt><dd><p>cellular barcode sequence and an optional suffix.
The sequence part is similar to the CR tag</p>
</dd>
<dt>CR:Z:sequence+  Cellular barcode. The uncorrected sequence bases of the</dt><dd><p>cellular barcode as reported by the sequencing machine,
with the corresponding base quality scores (optionally)
stored in CY.</p>
</dd>
<dt>CY:Z:qualities+ Phred quality of the cellular barcode sequence in CR tag</dt><dd><p>Same encoding as QUAL, i.e., Phred score + 33.</p>
</dd>
<dt>MI:Z:str        Molecular Identifier. A unique ID within the SAM file</dt><dd><p>for the source molecule from which this read is derived.</p>
</dd>
<dt>OX:Z:sequence+  Raw (uncorrected) unique molecular identifier bases,</dt><dd><p>with  quality scores (optionally) stored in the BZ tag.</p>
</dd>
<dt>BZ:Z:qualities+ Phred quality of the (uncorrected) unique molecular</dt><dd><p>identifier sequence in the OX tag.
Same encoding as QUAL, i.e., Phred score + 33.</p>
</dd>
<dt><a class="reference external" href="RX:Z:sequence+">RX:Z:sequence+</a>  Sequence bases from the unique molecular identifier.</dt><dd><p>These could be either corrected or uncorrected.
Unlike MI, the value may be non-unique in the file.</p>
</dd>
<dt>QX:Z:qualities+ Phred quality of the unique molecular identifier</dt><dd><p>sequence in the RX tag.
Same encoding as QUAL, i.e., Phred score + 33</p>
</dd>
<dt>OA:Z:(RNAME,POS,strand,CIGAR,MAPQ,NM;)+ The original alignment</dt><dd><p>information of the record prior to realignment or
unalignment by a subsequent tool.</p>
</dd>
</dl>
<p>OC:Z:cigar      Original CIGAR, usually before realignment.</p>
<dl class="simple">
<dt>CT:Z:strand ;type (;key (=value )?)* Complete read annotation tag</dt><dd><p>used for consensus annotation dummy features.</p>
</dd>
<dt>PT:Z:annotag(|annotag)* where each annotag matches</dt><dd><p>start;end;strand;type(;key(=value)?)*
Read annotations for parts of the padded read sequence.</p>
</dd>
</dl>
</div></blockquote>
<p>see <a class="reference external" href="https://samtools.github.io/hts-specs/SAMtags.pdf">https://samtools.github.io/hts-specs/SAMtags.pdf</a></p>
<p>Recommended try accept for use on raw alignment with fall back
to calling on only the tag byte string.</p>
<p>Please test carefully on your BAM output as in complicated output the
regular expression based extraction of the tag can be error prone</p>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">xenomapper2</a><ul>
<li><a class="reference internal" href="#xenomapper2-xenomapper2-module">xenomapper2.xenomapper2 module</a></li>
<li><a class="reference internal" href="#xenomapper2-cli-module">xenomapper2.cli module</a></li>
<li><a class="reference internal" href="#pylazybam-tags-module">pylazybam.tags module</a></li>
</ul>
</li>
</ul>

  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/xenomapper2.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">xenomapper2 2.0.0a1 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2011-2020, Matthew J Wakefield.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.4.1.
    </div>
  </body>
</html>